### PR TITLE
Add user assigned MSI support

### PR DESCRIPTION
**Pull request by kkmsft**
**What this PR does / why we need it**:
Adds the support for generating tokens via user assigned MSI. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # 

**Special notes for your reviewer**:

**Release note**:

```release-note
Add support for using User Assigned MSI (https://docs.microsoft.com/en-us/azure/active-directory/managed-service-identity/overview) with Kubernetes cluster on Azure.
```

### CODE REVIEW

**Comment by kkmsft** [comment]
cc @khenidak

File: pkg/cloudprovider/providers/azure/auth/azure_auth.go [hunk 1]
```diff
@@ -44,28 +44,39 @@ type AzureAuthConfig struct {
 	AADClientCertPassword string `json:"aadClientCertPassword" yaml:"aadClientCertPassword"`
 	// Use managed service identity for the virtual machine to access Azure ARM APIs
 	UseManagedIdentityExtension bool `json:"useManagedIdentityExtension" yaml:"useManagedIdentityExtension"`
+	// UserAssignedMSI contains the Client ID of the user assigned MSI which is assigned to the underlying VMs. If empty the user assigned identity is not used.
+	// More details of the user assigned identity can be found at: https://docs.microsoft.com/en-us/azure/active-directory/managed-service-identity/overview
+	// For the user assigned identity specified here to be used, the UseManagedIdentityExtension has to be set to true.
+	UserAssignedMSI string `json:"userAssignedMSI" yaml:"userAssignedMSI"`
```
Comment (feiskyer) [inline pkg/cloudprovider/providers/azure/auth/azure_auth.go]: use `UserAssignedIdentity` instead?  UserAssignedMSI seems overlapping with UseManagedIdentityExtension.

Comment (feiskyer) [inline pkg/cloudprovider/providers/azure/auth/azure_auth.go]: use UserAssignedIdentityID for clear? UserAssignedMSI seems overlapping with UseManagedIdentityExtension.

File: pkg/cloudprovider/providers/azure/auth/azure_auth.go [hunk 2]
```diff
@@ -44,28 +44,39 @@ type AzureAuthConfig struct {
...
-	}
-
 	if config.UseManagedIdentityExtension {
 		glog.V(2).Infoln("azure: using managed identity extension to retrieve access token")
 		msiEndpoint, err := adal.GetMSIVMEndpoint()
 		if err != nil {
 			return nil, fmt.Errorf("Getting the managed service identity endpoint: %v", err)
 		}
+		if len(config.UserAssignedMSI) > 0 {
+			glog.V(2).Info("azure: using User Assigned MSI to retrieve access token")
```
Comment (khenidak) [inline pkg/cloudprovider/providers/azure/auth/azure_auth.go]: nit: this is a bit too high verbosity. most clusters run with `2` -- can we make it `4`?

File: pkg/cloudprovider/providers/azure/auth/azure_auth.go [hunk 3]
```diff
@@ -44,28 +44,39 @@ type AzureAuthConfig struct {
...
+			return adal.NewServicePrincipalTokenFromMSIWithUserAssignedID(msiEndpoint,
+				env.ServiceManagementEndpoint,
+				config.UserAssignedMSI)
+		}
+		glog.V(2).Info("azure: using System Assigned MSI to retrieve access token")
```
Comment (khenidak) [inline pkg/cloudprovider/providers/azure/auth/azure_auth.go]: same as above. let us use `4`

**Review by khenidak — CHANGES_REQUESTED** [review]

**Comment by khenidak** [comment]
Let us remove the `wip` tag and prep this for merge. Thanks a lot !

**Review by feiskyer — COMMENTED** [review]
/lgtm

**Pull request merged**